### PR TITLE
Support all operation types in User Defined Index (UDI) interface

### PR DESCRIPTION
**Pull request by zaidoon1**
Remove the restriction that limited UDI to ingest-only, Puts-only use cases. This enables UDI plugins (including the trie index from #14310) to work with all operation types: Put, Delete, Merge, SingleDelete, PutEntity, etc.

Part of https://github.com/facebook/rocksdb/issues/12396

### CODE REVIEW

**Comment by zaidoon1** [comment]
some notes: I don't have db level tests, currently i'm ussing the sstfilewriter for tests which doesn't support all operations so that's not great. 

also we are missing things like seekToFirst, etc.. so i'll see if i can add that too as part of this

**Comment by zaidoon1** [comment]
- [ ] 

> some notes: I don't have db level tests, currently i'm ussing the sstfilewriter for tests which doesn't support all operations so that's not great.
> 
> also we are missing things like seekToFirst, etc.. so i'll see if i can add that too as part of this

this is done for now. I think what i have here is good for this PR.

**Comment by zaidoon1** [comment]
FYI, still working on some bugs i found. please ignore this PR for now.

**Comment by zaidoon1** [comment]
the current changes look good to me. I've run a 10 minute blackbox crash test without any issues. Currently running a 1 hour blackbox crash test then 1 hour whitebox crash test to really make sure we are good

File: utilities/trie_index/trie_index_factory.cc [hunk 1]
```diff
@@ -77,20 +77,24 @@ Slice TrieIndexBuilder::AddIndexEntry(const Slice& last_key_in_current_block,
...
-    // all-0xFF keys like "\xff\xff" — no byte can be incremented). When
-    // this happens AND the previous entry has the same separator, the last
-    // block is actually part of a same-user-key run. Without this check,
-    // the last block would get seqno=kMaxSequenceNumber (sentinel), but
-    // Finish() would group it into the run and hit the assert that overflow
-    // blocks must have real seqnos.
+    // Last block: use the last key itself as the separator, NOT a shortened
+    // successor. The standard index builder (ShortenedIndexBuilder) only
+    // calls FindShortInternalKeySuccessor when shortening_mode ==
+    // kShortenSeparatorsAndSuccessor, which is not the default. With the
```
Comment (zaidoon1) [inline utilities/trie_index/trie_index_factory.cc:84]: this was causing the crash tests to fail. I'm not sure what we want to do in general and whether we want to pass the block options that sets the mode and change behaviour depending on that

**Comment by zaidoon1** [comment]
@xingbowang this is now ready for review whenever you have a chance.

File: tools/db_crashtest.py [hunk 2]
```diff
@@ -892,33 +892,14 @@ def finalize_and_sanitize(src_params):
...
-        dest_params["delrangepercent"] = 0
-        # Ingestion with standalone range deletions is incompatible
-        dest_params["test_ingest_standalone_range_deletion_one_in"] = 0
         # Parallel compression is incompatible with UDI
         dest_params["compression_parallel_threads"] = 1
-        # Trie UDI has a known issue with prefix scanning where certain prefix
-        # patterns cause "SeekToFirst not supported" errors. Disable prefix
-        # scanning and redistribute its percentage to reads.
-        dest_params["readpercent"] += dest_params.get("prefixpercent", 0)
-        dest_params["prefixpercent"] = 0
```
Comment (xingbowang) [inline tools/db_crashtest.py:935]: We are removing quite a lot of these. Could you update the PR with stress test results that have these configuration on? These configuration could be overridden in the db_crashtest.py command line. I want to make sure we have these well tested before removing them. It caused bunch of stress test internally before which created lots of work for us to identify the root cause and add these configs.

**Comment by xingbowang** [comment]
M1 — key_type_log_ thread safety (2 agents agree): The TestUserDefinedIndexFactory::key_type_log_ shared vector has no synchronization. Not a production bug, but a latent data race in test infrastructure.

M2 — SeekToFirstAndGetResult default assumes BytewiseComparator (2 agents agree): The default implementation SeekAndGetResult(Slice()) is incorrect for non-bytewise comparators. Current risk is low (trie UDI overrides this and only supports BytewiseComparator), but third-party UDI implementations could hit silent bugs. Consider making it pure-virtual.

M3 — Simultaneous removal of all stress test restrictions (2 agents agree): All operation-type restrictions removed at once. The targeted DB-level tests (22 test cases) are thorough, but the crash test explores a much wider combinatorial space. Acceptable for EXPERIMENTAL, but worth monitoring.

M4 — prev_key_scratch_ string copy in NextAndGetResult: Could use std::move instead of copy to avoid per-Next() heap allocation for keys exceeding SSO threshold. One-line change, zero risk.

M5 — Missing degenerate SST edge cases at DB level: No DB-level tests for single-entry SST, deletion-only SST, or all-same-key SST. SST-level tests exist but don't exercise the full DB path.

File: utilities/trie_index/trie_index_factory.cc (hunk 1 above)
Comment (xingbowang) [reply utilities/trie_index/trie_index_factory.cc:84]: Could you elaborate more on the issue?

File: table/block_based/user_defined_index_wrapper.h [hunk 3]
```diff
@@ -260,8 +295,10 @@ class UserDefinedIndexIteratorWrapper
       valid_ = result_.bound_check_result == IterBoundCheck::kInbound;
       if (valid_) {
         ikey_.Set(result_.key, 0, ValueType::kTypeValue);
```
Comment (xingbowang) [inline table/block_based/user_defined_index_wrapper.h]: Could you add a comment to explain this? Essentially set result with internal key using user key. This is a general for multiple other APIs. Maybe extract a function helper to make this more explicit.

File: utilities/trie_index/trie_index_factory.cc (hunk 1 above)
Comment (zaidoon1) [reply utilities/trie_index/trie_index_factory.cc:84]: The trie index builder was unconditionally calling `FindShortSuccessor()` on the last block's separator key, but the standard `ShortenedIndexBuilder::GetSeparatorWithSeq` only does this when `shortening_mode == kShortenSeparatorsAndSuccessor`. With the default `kShortenSeparators` mode, the standard builder uses `last_key_in_current_block` as-is for the last block (see index_builder.h lines 278-286). This mismatch meant the trie's last-block separator covered a wider key range than the actual data, for example, if the last key was "9\xff\xff", `FindShortSuccessor` would produce ":" (0x3A), making the trie claim it covers keys up to ":". When a seek target fell between the real last key and the shortened successor, the trie would return a valid block that contained no matching data, which caused prefix scan iterators to desynchronize in test_batches_snapshots with deletes enabled. The fix here simply makes the trie match the standard builder's default behavior: use the last key directly, no shortening. 

My question was about  whether the trie should support all three shortening modes.

Right now, I believe the answer is no. The trie and standard index are used independently (never for the same query), so they don't need to produce identical separators. Using the last key as-is is the safest choice: it's correct regardless of the configured mode, and the alternative (FindShortSuccessor) only saves a few bytes on one trie entry while introducing the risk of claiming a wider range than the data supports. But would like your thought on this

**Comment by zaidoon1** [comment]
> M1 — key_type_log_ thread safety (2 agents agree): The TestUserDefinedIndexFactory::key_type_log_ shared vector has no synchronization. Not a production bug, but a latent data race in test infrastructure.
> 
> M2 — SeekToFirstAndGetResult default assumes BytewiseComparator (2 agents agree): The default implementation SeekAndGetResult(Slice()) is incorrect for non-bytewise comparators. Current risk is low (trie UDI overrides this and only supports BytewiseComparator), but third-party UDI implementations could hit silent bugs. Consider making it pure-virtual.
> 
> M3 — Simultaneous removal of all stress test restrictions (2 agents agree): All operation-type restrictions removed at once. The targeted DB-level tests (22 test cases) are thorough, but the crash test explores a much wider combinatorial space. Acceptable for EXPERIMENTAL, but worth monitoring.
> 
> M4 — prev_key_scratch_ string copy in NextAndGetResult: Could use std::move instead of copy to avoid per-Next() heap allocation for keys exceeding SSO threshold. One-line change, zero risk.
> 
> M5 — Missing degenerate SST edge cases at DB level: No DB-level tests for single-entry SST, deletion-Nonly SST, or all-same-key SST. SST-level tests exist but don't exercise the full DB path.


M1: not a bug, it's test only code and all reads happen after synchronous Flush/CompactRange with auto-compactions disabled. No concurrent access.

M2: The trie UDI overrides this (line 238-267 of trie_index_factory.cc). The default has a documented warning. Making it pure virtual would break backward compatibility.

M4: copy is actually necessary. We can't use std::move because current_key_scratch_ is reused in the overflow path (line 405: result->key = Slice(current_key_scratch_)). Keys are typically small (within SSO), so the copy is cheap.

File: tools/db_crashtest.py (hunk 2 above)
Comment (zaidoon1) [reply tools/db_crashtest.py:935]: Here are the stress test results for all previously-restricted configurations. Each test was run with `--use_trie_index=1` and 8 threads for 120 seconds (terminated by SIGALRM). Tests that should fail are also included to demonstrate the test harness catches real incompatibilities.

### Previously-Restricted Features — Now Passing

All of the following were previously blocked by `db_crashtest.py` overrides or `db_stress_tool.cc` hard errors when `use_trie_index=1`. They now pass:

| Test | Key Flags | Duration | Threads | Max Ops/Thread | Result |
|------|-----------|----------|---------|----------------|--------|
| **Comprehensive** | `iterpercent=15 prefixpercent=10 delpercent=10 delrangepercent=5 use_merge=1 use_put_entity_one_in=10` | 120s | 8 | 400,000+ | **PASS** |
| **Integrated BlobDB** | `enable_blob_files=1 min_blob_size=0 enable_blob_garbage_collection=1 blob_garbage_collection_age_cutoff=0.5` + iter/prefix/merge/PutEntity | 120s | 8 | 400,000+ | **PASS** |
| **Optimistic TransactionDB** | `use_txn=1 use_optimistic_txn=1` + iter/prefix/merge/PutEntity/delete | 120s | 8 | 400,000+ | **PASS** |
| **Pessimistic TransactionDB** | `use_txn=1 use_optimistic_txn=0` + iter/prefix/merge/PutEntity/delete | 120s | 8 | 60,000+ | **PASS** |

<details>
<summary>Comprehensive test command</summary>

```bash
./db_stress \
  --use_trie_index=1 --mmap_read=0 --compression_parallel_threads=1 \
  --remote_compaction_worker_threads=0 --test_batches_snapshots=0 \
  --iterpercent=15 --prefixpercent=10 --readpercent=25 --writepercent=25 \
  --delpercent=10 --delrangepercent=5 --customopspercent=10 \
  --use_merge=1 --use_put_entity_one_in=10 \
  --prefix_size=7 --nooverwritepercent=1 --disable_wal=1 \
  --max_key=100000 --max_key_len=3 --column_families=1 \
  --destroy_db_initially=1 --reopen=0 --ops_per_thread=100000000 --threads=8 \
  --expected_values_dir=<dir> --value_size_mult=32 \
  --flush_one_in=1000 --compact_range_one_in=1000000 --verify_db_one_in=10000 \
  --index_type=0 --block_size=16384 --compression_type=none \
  --clear_column_family_one_in=0 --read_fault_one_in=0 \
  --metadata_read_fault_one_in=0 --metadata_write_fault_one_in=0
# Exit code: 142 (SIGALRM — ran full 120s)
```
</details>

<details>
<summary>BlobDB test command</summary>

```bash
./db_stress \
  --use_trie_index=1 --mmap_read=0 --compression_parallel_threads=1 \
  --remote_compaction_worker_threads=0 --test_batches_snapshots=0 \
  --enable_blob_files=1 --min_blob_size=0 --blob_file_size=1048576 \
  --enable_blob_garbage_collection=1 --blob_garbage_collection_age_cutoff=0.5 \
  --iterpercent=10 --prefixpercent=5 --readpercent=35 --writepercent=30 \
  --delpercent=10 --delrangepercent=0 --customopspercent=10 \
  --use_merge=1 --use_put_entity_one_in=10 \
  --prefix_size=7 --nooverwritepercent=1 --disable_wal=1 \
  --max_key=100000 --max_key_len=3 --column_families=1 \
  --destroy_db_initially=1 --reopen=0 --ops_per_thread=100000000 --threads=8 \
  --expected_values_dir=<dir> --value_size_mult=32 \
  --flush_one_in=1000 --compact_range_one_in=1000000 --verify_db_one_in=10000 \
  --index_type=0 --block_size=16384 --compression_type=none \
  --clear_column_family_one_in=0 --read_fault_one_in=0 \
  --metadata_read_fault_one_in=0 --metadata_write_fault_one_in=0
# Exit code: 142 (SIGALRM — ran full 120s)
```
</details>

<details>
<summary>Transaction test commands</summary>

```bash
# Optimistic TransactionDB
./db_stress \
  --use_trie_index=1 --mmap_read=0 --compression_parallel_threads=1 \
  --remote_compaction_worker_threads=0 --test_batches_snapshots=0 \
  --use_txn=1 --use_optimistic_txn=1 \
  --iterpercent=10 --prefixpercent=5 --readpercent=35 --writepercent=30 \
  --delpercent=10 --delrangepercent=0 --customopspercent=10 \
  --use_merge=1 --use_put_entity_one_in=10 \
  --prefix_size=7 --nooverwritepercent=1 --disable_wal=1 \
  --max_key=100000 --max_key_len=3 --column_families=1 \
  --destroy_db_initially=1 --reopen=0 --ops_per_thread=100000000 --threads=8 \
  --expected_values_dir=<dir> --value_size_mult=32 \
  --flush_one_in=1000 --compact_range_one_in=1000000 --verify_db_one_in=10000 \
  --index_type=0 --block_size=16384 --compression_type=none \
  --clear_column_family_one_in=0 --read_fault_one_in=0 \
  --metadata_read_fault_one_in=0 --metadata_write_fault_one_in=0
# Exit code: 142 (SIGALRM — ran full 120s)

# Pessimistic TransactionDB (same but --use_optimistic_txn=0)
# Exit code: 142 (SIGALRM — ran full 120s)
```
</details>

### Remaining Restrictions — Correctly Rejected

These features are still incompatible with trie UDI and are correctly caught:

| Test | Flag | Result | Error Message |
|------|------|--------|---------------|
| **mmap_read** | `--use_trie_index=1 --mmap_read=1` | **FAIL (exit 1)** | `Error: use_trie_index is incompatible with mmap_read. The trie index uses zero-copy pointers into block data which is unsafe with mmap'd reads.` |
| **Backward scan** | `--use_trie_index=1 --test_backward_scan=1` | **Auto-disabled** | `db_stress_tool.cc` silently sets `test_backward_scan=false` (SeekToLast/SeekForPrev/Prev not supported by trie UDI) |

### End-to-End: `db_crashtest.py` Blackbox Test

Also ran the full crash test framework with `--use_trie_index=1` forced on:

```bash
python3 tools/db_crashtest.py blackbox --use_trie_index=1 --duration=60 --interval=30 --simple
# Exit code: 0
```

This ran 3 iterations of `db_stress` with fully randomized parameters (different compaction styles, compression, fault injection, remote compaction, etc.) — all with `use_trie_index=1`. The crash test framework correctly applied `mmap_read=0` and `compression_parallel_threads=1` via `finalize_and_sanitize()`. The final verification-only pass confirmed data integrity:

```
Crash-recovery verification passed :)
Verification successful
```

Comment (zaidoon1) [reply tools/db_crashtest.py:935]: I'm running a longer test right now and will post the results, also please let me know if i've run all the commands correctly/addressed the concern or if I need to run some other commands too

Comment (zaidoon1) [reply tools/db_crashtest.py:935]: Updated with 10-minute runs (all 4 tests ran in parallel, 8 threads each):

| Test | Key Flags | Duration | Threads | Max Ops/Thread | Result |
|------|-----------|----------|---------|----------------|--------|
| **Comprehensive** | `iterpercent=15 prefixpercent=10 delpercent=10 delrangepercent=5 use_merge=1 use_put_entity_one_in=10` | 10 min | 8 | 1,300,000+ | **PASS** |
| **Integrated BlobDB** | `enable_blob_files=1 min_blob_size=0 enable_blob_garbage_collection=1 blob_garbage_collection_age_cutoff=0.5` + iter/prefix/merge/PutEntity | 10 min | 8 | 400,000+ | **PASS** |
| **Optimistic TransactionDB** | `use_txn=1 use_optimistic_txn=1` + iter/prefix/merge/PutEntity/delete | 10 min | 8 | 1,300,000+ | **PASS** |
| **Pessimistic TransactionDB** | `use_txn=1 use_optimistic_txn=0` + iter/prefix/merge/PutEntity/delete | 10 min | 8 | 400,000+ | **PASS** |

All 4 exited with code 142 (SIGALRM — ran the full 10 minutes without any crash, assertion failure, or verification error). Same commands as above, just with a 600s alarm instead of 120s.

File: table/block_based/user_defined_index_wrapper.h (hunk 3 above)
Comment (zaidoon1) [reply table/block_based/user_defined_index_wrapper.h]: done

File: utilities/trie_index/trie_index_factory.cc (hunk 1 above)
Comment (xingbowang) [reply utilities/trie_index/trie_index_factory.cc:84]: I think what you have is good for now. Trie index will be an alternative index implementation optimized. We should be good.

File: tools/db_crashtest.py (hunk 2 above)
Comment (xingbowang) [reply tools/db_crashtest.py:935]: Thanks for posting the result. Did you double check all those options are actually enabled in the run. Sometimes, some of the options disable other options, so it is better to double check this. Thank you.

**Comment by xingbowang** [comment]
Another feedback. Please avoid rebase or squash as much as possible, it is hard for me to read what was changed since last time. Thank you.

**Comment by xingbowang** [comment]
I see you have added a lot more unit test coverage, which is great. Based on my past experience, one of the places that could be prone to bug is multiple levels of SST files combined with delete range. Could you add/extend existing randomized test for that?

File: utilities/trie_index/trie_index_test.cc [hunk 4]
```diff
@@ -5169,6 +5369,524 @@ TEST_F(TrieSeekBenchmark, TrieVsRealIndexBlockIter) {
   fprintf(stderr, "\n");
 }
 
+// ============================================================================
+// Mixed key type tests — verifies UDI works with Delete, Merge, etc.
+// ============================================================================
+
+TEST_F(TrieIndexSSTTest, MixedKeyTypesWithTrieUDI) {
```
Comment (xingbowang) [inline utilities/trie_index/trie_index_test.cc:4662]: The test is already more than 5000+ lines of code. Any opportunity to dedup some of the logic such as similar boiler plate set up or use parameterized test? Or use fuzzy test to cover some common pattern?

File: utilities/trie_index/trie_index_db_test.cc [hunk 5]
```diff
@@ -0,0 +1,2821 @@
...
+  // Verifies Get with a snapshot through both indexes.
+  void VerifyGetBothIndexes(const Snapshot* snap, const std::string& key,
+                            const std::string& expected_value) {
+    for (auto base_ro : {StandardIndexReadOptions(), TrieIndexReadOptions()}) {
+      SCOPED_TRACE(base_ro.table_index_factory ? "trie index"
+                                               : "standard index");
+      base_ro.snapshot = snap;
+      std::string value;
+      ASSERT_OK(db_->Get(base_ro, key, &value));
+      ASSERT_EQ(value, expected_value);
```
Comment (xingbowang) [inline utilities/trie_index/trie_index_db_test.cc:185]: Test will not be properly stopped. 
Wrap the call site with ASSERT_NO_FATAL_FAILURE

**Comment by xingbowang** [comment]
I think we are getting very close. After addressing the comments above, we should be good.

**Comment by xingbowang** [comment]
Looks like some crash test failure with too many open files. Not sure it is related with this change or not.

File: utilities/trie_index/trie_index_db_test.cc (hunk 5 above)
Comment (zaidoon1) [reply utilities/trie_index/trie_index_db_test.cc:185]: done

File: utilities/trie_index/trie_index_test.cc (hunk 4 above)
Comment (zaidoon1) [reply utilities/trie_index/trie_index_test.cc:4662]: i've refactored the code to dedup as much as possible

**Comment by zaidoon1** [comment]
> I see you have added a lot more unit test coverage, which is great. Based on my past experience, one of the places that could be prone to bug is multiple levels of SST files combined with delete range. Could you add/extend existing randomized test for that?

done, let me know if i missed any cases or should add more.

**Comment by zaidoon1** [comment]
> Looks like some crash test failure with too many open files. Not sure it is related with this change or not.

double checked this and it seems to be some issue with ci or an existing issue, at least it's not failing anywhere close to the trie/UDI code.

**Comment by xingbowang** [comment]
LGTM, thanks for making this huge improvement!

**Comment by xingbowang** [comment]
Would you mind to do a final merge with upstream/main? It will make it easier to pass CI

**Comment by zaidoon1** [comment]
> Would you mind to do a final merge with upstream/main? It will make it easier to pass CI

done
